### PR TITLE
Add a pylibcudf regex search example

### LINKED ISSUE #21078: [FEA] write a grep utility with cuDF
**Is your feature request related to a problem? Please describe.**
Regex features are available in cudf string APIs and nvtext APIs. 
`grep` is widely used utility. It would cool to try creating `cugrep` with cudf/nvtext and document the throughput.

**Describe the solution you'd like**
Create `cugrep` utility in our python codebase or a C++ example with libcudf APIs.

**Additional context**
Benchmark it without GDS (most common usecase).
Benchmark on GPU in PCIE, and on Grace systems with fast C2C.

### PR DESCRIPTION
**Pull request by JayYarlagadda**
## Description

This reframes #21078 as a focused pylibcudf example rather than a packaged grep-compatible CLI.

The example:

- reads a plain-text file into one GPU string per line with `pylibcudf.io.text.multibyte_split`;
- compiles and runs a libcudf regular expression, including optional ignore-case matching;
- validates a trusted expected match count when provided;
- reports synchronized end-to-end and resident-data scan throughput; and
- can compare POSIX and GDS paths in isolated child processes so KvikIO is initialized independently.

GDS-on results are reported only when cuFile says native GDS is available and compatibility fallback is disabled. This prevents a successful cuFile compatibility-mode read from being mislabeled as native GDS performance.

The implementation lives in `python/pylibcudf/examples/regex_search.py`, with usage and benchmarking notes in the adjacent README.

Closes #21078.

### Native GDS

Karthikeyan tested the example on one NVIDIA B200 with `pylibcudf 26.12.0a111` and a GDS-enabled ext4 filesystem. `gdsio` verified the storage path, `is_gds_available=True`, and `allow_compat_mode=False`.

GDS-on and GDS-off runs produced identical file sizes, line counts, and match counts across input sizes from 1 MiB through 8 GiB and across multiple regex patterns. The most repeatable large-input result was at 8 GiB, where native GDS improved end-to-end throughput by 15.2–17.7%. Full benchmark results and methodology are recorded in the PR discussion.

The native-GDS run also exposed two API issues that are fixed in `c35bc9b3`: cuFile is initialized before its runtime properties are read, and match counting uses `apply_retention_mask`.

### GDS-off

The no-GDS path was exercised on:

- Tesla T4, driver `580.159.04`;
- Python `3.12.13`;
- `pylibcudf 26.10.0a276.post260820171429`; and
- `libcudf 26.10.0a276.post260820171429`.

For a deterministic 1.248 GB log containing 16,000,000 lines:

| pattern | mode | expected/actual matches | end-to-end | resident scan |
|---|---|---:|---:|---:|
| `ERROR` | case-sensitive | 160,000 | 2.38 GB/s | 42.95 GB/s |
| `error` | ignore-case | 320,000 | 2.44 GB/s | 43.62 GB/s |

The benchmark synchronizes the CUDA stream around every timed region. Small-file, ignore-case, and terminal-newline correctness cases were also validated.

A separate A100 run with a current `pylibcudf 26.12` nightly passed all 8 focused tests, exercised the GDS-off path end to end, and confirmed that a host without native GDS rejects cuFile compatibility fallback instead of reporting it as native GDS.

Additional validation completed:

- pylibcudf Ruff formatting and lint checks;
- `git diff --check`;
- pre-commit checks; and
- 8 focused regression tests covering real GPU regex execution, boundary cases, timing, expected-match validation, and GDS fallback handling.

## Checklist

- [x] I am familiar with the Contributing Guidelines.
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by davidwendt** [comment]
@JayYarlagadda Is this ready for review? I see that the PR is still in Draft.

**Comment by beckernick** [comment]
How does this compare to ripgrep?

https://ripgrep.dev/benchmarks/

**Comment by JayYarlagadda** [comment]
Good shout on ripgrep — ran it on the same T4 to be fair (synthetic log, pattern `ERROR`, best of 3, no GDS, and all four tools returned the same match count):

| size | grep | ripgrep | cudfgrep e2e | cudfgrep scan-only |
|---|---|---|---|---|
| 1 GB | 1.16 | 2.35 | 2.20 | 21.79 |
| 2 GB | 1.17 | 2.42 | 2.27 | 18.86 |
| 4 GB | 1.16 | 2.43 | 2.06 | 19.07 |

(GB/s)

So yeah, ripgrep is roughly 2x GNU grep — a much fairer baseline, glad you flagged it. On a single cached file the end-to-end cudfgrep number lands about the same as ripgrep, since most of the time goes into reading the file and copying it to the GPU rather than the actual matching.

Where the GPU pulls ahead is the scan itself — once the data is resident it matches at ~20 GB/s, roughly 8-9x ripgrep. So it makes the most sense when the text is already on the device (as part of a cuDF pipeline), or when you're scanning the same data repeatedly / with multiple patterns.

One limitation I hit: it loads the whole file into VRAM, so 8 GB OOM'd on the 16 GB T4. `read_text` takes a `byte_range` though, so chunked reading (and overlapping the load with the compute) plus GDS is the obvious next step for bigger-than-VRAM files.

Happy to fold this into the PR description if it's useful.

**Comment by JayYarlagadda** [comment]
@davidwendt yep, marking it ready now — thanks for the nudge. Also dropped a ripgrep comparison above.

**Comment by davidwendt** [comment]
Just curious what version of cudf was used to produce the results. There was a significant performance improvement to the regex code merged into main on Tuesday afternoon (7/14). Are the numbers based on the current code (as of the last couple days)?

**Comment by JayYarlagadda** [comment]
Good question — these were on cudf 26.02.01. That's the newest version I could get running on Kaggle's free T4, since the 26.6 wheels want a newer CUDA driver than the free tier has. So yeah, they're from before the 7/14 regex changes — which means the scan-only numbers are probably on the low side compared to current main.

I don't have a box with a newer driver to rerun on main right now, but I'm happy to if you can point me at one, or I can give the devcontainer a shot. I'd expect the resident-scan throughput to only go up.

**Comment by JayYarlagadda** [comment]
Pushed a commit addressing the review feedback:

- Added the suggested tests — CLI flag forwarding (`-i`/`-v`/`-w`/`-x`), `-o -v` (no output, exit 1), `--benchmark` on stdin (exit 2), an empty file, and a file without a trailing newline.
- Filled out the `main()` docstring (argv, return value, exit codes).

On a few of the other suggestions:

- **`-o -v` exit code:** I checked GNU grep and it exits 1 with no output for `-o -v`, so the current behavior already matches — added a test to lock that in.
- **GDS / import order:** since the tool lives under the `cudf` namespace, the entry point imports `cudf` regardless, so there's no way to set the env var before cuDF loads from a pre-import shim. cuDF doesn't expose a runtime API for KvikIO compat mode either — it's the `KVIKIO_COMPAT_MODE` env var, which I set before the first read. Happy to change the approach if you'd prefer something else.
- **`-w` on non-word patterns (e.g. `@`):** fair edge case, but the clean fix needs lookbehind/lookahead, which I don't think libcudf's regex supports; kept `\b` (correct for word patterns) for now. Can follow up if it matters.
- **libcudf regex validator:** reasonable — currently an unsupported pattern still errors out (exit 2), just at the libcudf layer rather than a pre-check. Can add a validator as a follow-up.

**Comment by JayYarlagadda** [comment]
Thanks for the thorough review — pushed `f125467` covering most of it:

- Expanded the tests: CLI flag forwarding (`-i`/`-v`/`-w`/`-x`), `-o -v` (asserts no output + exit 1), `--benchmark` with stdin (exit 2), plus empty-file and no-trailing-newline cases.
- Filled out the `main()` docstring (argv, return value, and the 0/1/2 exit codes).

A few I'm holding off on, with reasoning:

- **GDS pre-import shim:** the toggle sets `KVIKIO_COMPAT_MODE`, which is cuDF's documented way to control GDS (there's no Python runtime API for it). Fair point that the entry point imports `cudf` before `main()` runs — before restructuring I want to confirm whether kvikio reads the var lazily at IO time or caches it at import, so I'll verify and move it into a pre-import shim only if it's actually needed.
- **`-w` on non-word patterns** (e.g. `@`): a real GNU-grep edge case, but it only affects patterns made entirely of non-word characters. I'd rather keep the common-case behavior and note this as a known limitation for a follow-up.
- **Dedicated libcudf regex validator:** `str.contains`/`findall` already raise on unsupported flags, so clearly-invalid patterns fail at the cuDF layer today. A standalone validator is reasonable but broader than this PR — happy to add it as a follow-up if you'd like.

**Comment by JayYarlagadda** [comment]
﻿Hi @karthikeyann — just checking in on this one. Is there anything you'd like me to adjust, or is it mainly waiting on a CI run at this point? I'd also be happy to re-run the benchmarks on current `main` if that would be useful, given the recent regex improvements.

**Comment by JayYarlagadda** [comment]
@davidwendt @karthikeyann — got a chance to rerun on current `main` (nightly `cudf 26.10.0a150`), and the recent regex work makes a real difference. Same T4, same synthetic log, pattern `ERROR`, match counts identical to grep:

| size | grep | ripgrep | cudfgrep e2e | cudfgrep scan-only |
|---|---|---|---|---|
| 1 GB | 1.15 | 2.36 | 2.54 | 103.6 |
| 2 GB | 1.13 | 2.41 | 2.44 | 90.8 |
| 4 GB | 1.15 | 2.41 | 2.40 | 93.0 |

(GB/s) Scan-only jumped from ~20 GB/s on the old 26.02 build to ~90–100 GB/s here — roughly 5×, and ~40× ripgrep on resident data. End-to-end is basically unchanged (~2.4 GB/s, still ~tied with ripgrep) since it's dominated by loading the file onto the GPU — so chunked reads + GDS remain the obvious lever there.

On that note: right now it loads the whole file into VRAM (8 GB OOMs on the 16 GB T4). I'm happy to add chunked reading via `read_text`'s `byte_range` — that should handle larger-than-VRAM inputs and let the load overlap with the scan. Glad to fold it into this PR or do it as a follow-up, whatever you'd prefer.

Happy to make any changes whenever someone gets a chance to look.

**Comment by JayYarlagadda** [comment]
@karthikeyann @davidwendt — just checking back one more time. Still glad to add the chunked-reading (`byte_range`) work if it'd be useful, or leave the PR as-is — whatever fits your priorities. No rush at all, and thanks for the time you've already put into this.

**Review by bdice — COMMENTED** [review]
I thought the intent of the issue proposing this was not to produce a serious CLI utility to compete with the many many regex searching tools out there, but just to have a demo of cuDF's regex processing capability and a general sense of the GPU performance.

I would think of this as a Python code example, not something we ship as a part of the `cudf` Python package as a built-in CLI utility.

Maybe this could be reimagined as a `pylibcudf` demo, because we definitely need more `pylibcudf` examples/demos. That would also reduce overhead, because the `cudf` pandas-like API is slower than `pylibcudf`.

tl;dr I don't think we should ship this as a CLI utility, moreso as demo code. It's great to see the performance numbers -- especially if we can get some tests with GDS -- but competing with tools like ripgrep is not the goal.

**Comment by JayYarlagadda** [comment]
@bdice Thanks, that clarifies the intended scope. I interpreted the earlier discussion about making this available through cuDF Python as a request for a shipped CLI, but I'm happy to reframe it as a focused demonstration instead.

I'll remove the package entry point and grep-compatibility surface, rewrite the core using pylibcudf, and update the benchmarks to include GDS enabled/disabled alongside end-to-end and resident-data scan timings. I'll treat ripgrep only as context rather than as something the example is intended to compete with.

I don't see an existing pylibcudf examples directory on main. Would python/pylibcudf/examples/regex_search.py be an appropriate location, or is there another examples/docs path you would prefer?

**Comment by JayYarlagadda** [comment]
@bdice @davidwendt — quick update while I’m working through the pylibcudf-demo rewrite.

I now have the focused demo running end to end on a current RAPIDS nightly:

- Tesla T4, driver `580.159.04`
- Python `3.12.13`
- `pylibcudf 26.10.0a276.post260820171429`
- `libcudf 26.10.0a276.post260820171429`
- 1.248 GB synthetic log, 16,000,000 lines

Correctness was checked using deterministic reference counts:

| pattern | mode | expected/actual matches | end-to-end | resident scan |
|---|---|---:|---:|---:|
| `ERROR` | case-sensitive | 160,000 | 2.38 GB/s | 42.95 GB/s |
| `error` | ignore-case | 320,000 | 2.44 GB/s | 43.62 GB/s |

I also checked the small-file and terminal-newline cases, including ensuring a final newline does not create a false empty-line match. The focused test suite currently has 36 passing tests and passes the cuDF Ruff configuration.

One correction to my earlier benchmark comment: the revised benchmark explicitly synchronizes the CUDA stream around every timed region. Please treat these ~43 GB/s resident-scan results as authoritative and the earlier ~90–100 GB/s numbers as superseded.

I also investigated GDS rather than treating a successful cuFile call as proof. Kaggle reports:

```text
is_gds_available: False
allow_compat_mode: True
nvidia-fs: not loaded
/dev/nvidia-fs: unavailable
```

`/kaggle/working` is an ext4 loop device, so cuFile was using its internal compatibility fallback. I’m therefore not reporting that as a GDS result. The revised demo detects this condition and rejects `--gds-mode on` with a clear error instead of publishing a misleading comparison.

Would either of you have access to a GDS-capable runner where I could run the same reproducible benchmark, or would you be willing to run it internally? I can provide the exact script and deterministic input-generation command. Otherwise, I’ll document native GDS as unverified and include the validated no-GDS results above.

For placement, unless you prefer another location, I’ll proceed with `python/pylibcudf/examples/regex_search.py` and keep it as a standalone demonstration rather than an installed CLI.

**Comment by JayYarlagadda** [comment]
The pylibcudf rewrite is now pushed in `7923490` and I’ve returned the PR to Draft while the new scope is reviewed.

The effective diff now contains only:

- `python/pylibcudf/examples/regex_search.py`
- `python/pylibcudf/examples/README.md`
- `python/pylibcudf/tests/test_regex_search_example.py`

The packaged `cudfgrep` entry point and grep-compatibility implementation are removed. GDS-off is the safe default; requested GDS comparisons run in isolated processes and require both native GDS availability and cuFile compatibility fallback to be disabled.

Local formatting, Ruff, diff checks, and the focused regression suite pass. I’ll leave this as Draft pending feedback on example placement and access to a native-GDS runner.

File: python/pylibcudf/examples/regex_search.py [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+    return plc.strings.contains.contains_re(lines, program, stream=stream)
+
+
+def count_matches(lines: Any, mask: Any, plc: Any, stream: Any) -> int:
+    """Count selected rows without converting the strings column to pandas."""
+    selected = plc.stream_compaction.apply_boolean_mask(
+        plc.Table([lines]), mask, stream=stream
+    )
+    stream.synchronize()
+    return selected.num_rows()
```
[the remark this answers is not in the document]
Comment (JayYarlagadda) [reply python/pylibcudf/examples/regex_search.py:150]: Addressed in c7e767f. count_matches now masks a zero-column Table([], num_rows=lines.size()), so it preserves the selected-row count without materializing matching strings. Verified with 37 local regression tests plus Ruff lint/format checks.

File: python/pylibcudf/tests/test_regex_search_example.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        repeats=2,
+        plc=plc,
+        stream=DEFAULT_STREAM,
+    )
+
+    assert result.bytes == path.stat().st_size
+    assert result.lines == 3
+    assert result.matches == expected_matches
+    assert result.end_to_end_seconds > 0
+    assert result.scan_seconds > 0
```
[the remark this answers is not in the document]
Comment (JayYarlagadda) [reply python/pylibcudf/tests/test_regex_search_example.py]: Addressed in c7e767f. test_benchmark_file now parameterizes empty-file, single-line, and final-line-without-newline cases while retaining byte and synchronized timing assertions. Test functions also have docstrings to satisfy coverage guidance. Verified with 37 local regression tests plus Ruff lint/format checks.

**Comment by karthikeyann** [comment]
Tested this on B200 with GDS (`gdsio` verified with 65MiB write/read at 3.63/3.68 GiB/s)
I had to add fixes to move from deprecated `apply_boolean_mask` to `apply_retention_mask`. 
# cudfgrep GDS benchmark results

Benchmark of the `pylibcudf` regex-search example from cuDF PR #23286.

## Environment

- GPU: one NVIDIA B200 (`CUDA_VISIBLE_DEVICES=0`)
- pylibcudf: `26.12.0a111`
- Native GDS validation:
  - `is_gds_available=True`
  - `allow_compat_mode=False`
- Input filesystem: `/raid` (ext4)
- Input format: deterministic 128-byte newline-terminated records
- Timing: best synchronized result from seven repetitions, except the 1 GiB
  case, which used five repetitions
- Each run included a correctness/warm-up pass before timing

All GDS-on and GDS-off runs produced identical file sizes, line counts, and
match counts.

## Input-size sweep

Regex pattern: `ERROR`

| Input size | Lines | Matches | GDS off end-to-end | GDS on end-to-end | Change |
|---:|---:|---:|---:|---:|---:|
| 1 MiB | 8,192 | 2 | 0.276 GB/s | 0.281 GB/s | +1.9% |
| 16 MiB | 131,072 | 32 | 2.907 GB/s | 2.876 GB/s | -1.0% |
| 64 MiB | 524,288 | 128 | 5.371 GB/s | 5.985 GB/s | +11.4% |
| 256 MiB | 2,097,152 | 512 | 4.333 GB/s | 4.240 GB/s | -2.1% |
| 1 GiB | 8,388,608 | 2,048 | 4.073 GB/s | 5.447 GB/s | **+33.7%** |

### Scan-only throughput

| Input size | GDS off | GDS on |
|---:|---:|---:|
| 1 MiB | 12.20 GB/s | 11.91 GB/s |
| 16 MiB | 156.53 GB/s | 140.96 GB/s |
| 64 MiB | 332.42 GB/s | 335.89 GB/s |
| 256 MiB | 333.85 GB/s | 332.01 GB/s |
| 1 GiB | 470.15 GB/s | 468.71 GB/s |

Scan-only throughput is effectively independent of GDS because the text is
already resident in GPU memory during that measurement.

## Larger input-size sweep

The same `ERROR` workload was extended to 2, 4, and 8 GiB. Each mode used the
best synchronized result from five repetitions. Two trials were collected per
size with the mode order reversed in the second trial to expose ordering and
page-cache variability.

| Input size | Lines | Matches | Trial | GDS off | GDS on | Change |
|---:|---:|---:|---:|---:|---:|---:|
| 2 GiB | 16,777,216 | 4,096 | off then on | 3.990 GB/s | 5.127 GB/s | +28.5% |
| 2 GiB | 16,777,216 | 4,096 | on then off | 5.192 GB/s | 5.291 GB/s | +1.9% |
| 4 GiB | 33,554,432 | 8,192 | off then on | 5.412 GB/s | 5.251 GB/s | -3.0% |
| 4 GiB | 33,554,432 | 8,192 | on then off | 3.860 GB/s | 5.134 GB/s | +33.0% |
| 8 GiB | 67,108,864 | 16,384 | off then on | 4.262 GB/s | 5.015 GB/s | +17.7% |
| 8 GiB | 67,108,864 | 16,384 | on then off | 4.258 GB/s | 4.905 GB/s | +15.2% |

Observed end-to-end ranges across the two trials were:

| Input size | GDS off range | GDS on range |
|---:|---:|---:|
| 2 GiB | 3.990-5.192 GB/s | 5.127-5.291 GB/s |
| 4 GiB | 3.860-5.412 GB/s | 5.134-5.251 GB/s |
| 8 GiB | 4.258-4.262 GB/s | 4.905-5.015 GB/s |

The 8 GiB result was the most repeatable: native GDS improved end-to-end
throughput by 15.2-17.7%. The GDS measurements were also steadier than the
non-GDS measurements across all three larger sizes. The 2 and 4 GiB non-GDS
results show enough page-cache and run-order variability that a single-trial
percentage would be misleading.

## Regex-parameter sweep

Input size: 256 MiB (2,097,152 lines)

| Pattern/options | Matches | GDS off end-to-end | GDS on end-to-end | Change |
|---|---:|---:|---:|---:|
| `ERROR` | 512 | 4.258 GB/s | 4.396 GB/s | +3.2% |
| `^(ERROR\|WARN)` | 1,024 | 3.687 GB/s | 4.204 GB/s | +14.0% |
| `code=[45][0-9][0-9]` | 1,536 | 4.262 GB/s | 5.565 GB/s | +30.6% |
| `error --ignore-case` | 1,024 | 4.412 GB/s | 5.364 GB/s | +21.6% |

## Interpretation

Native GDS operated correctly for every tested input size and regex option.
The clearest end-to-end benefit appeared at 1 GiB, where GDS improved measured
throughput by 33.7% in the initial sweep. In the extended sweep, the repeatable
8 GiB improvement was 15.2-17.7%. Smaller inputs are more sensitive to fixed
allocation, startup, page-cache, and I/O variance, so their individual
differences should not be treated as stable storage-performance claims.

This is an example-level benchmark rather than a raw storage benchmark. Its
correctness pass warms the workload, and the reported value is the best of the
timed repetitions. The non-GDS path may also benefit from the operating-system
page cache.

File: python/pylibcudf/examples/regex_search.py [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+
+def _require_native_gds() -> None:
+    """Raise when cuFile would use its internal compatibility fallback."""
+    try:
+        from kvikio import cufile_driver
+    except ImportError as exc:
+        raise RuntimeError(
+            "native GDS validation requires the kvikio Python package"
+        ) from exc
+    if not cufile_driver.get("is_gds_available"):
```
Comment (karthikeyann) [inline python/pylibcudf/examples/regex_search.py:76]: HIGH: Initialize cuFile before reading its runtime properties. On the tested native-GDS host, `get()` returned placeholder values until initialization, causing working GDS to be rejected.

```suggestion
    cufile_driver.initialize()
    if not cufile_driver.get("is_gds_available"):
```

File: python/pylibcudf/examples/regex_search.py [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+
+
+def scan_lines(lines: Any, program: Any, plc: Any, stream: Any) -> Any:
+    """Return a Boolean column indicating which lines match ``program``."""
+    return plc.strings.contains.contains_re(lines, program, stream=stream)
+
+
+def count_matches(lines: Any, mask: Any, plc: Any, stream: Any) -> int:
+    """Count selected rows without materializing the matching strings."""
+    selected = plc.stream_compaction.apply_boolean_mask(
```
Comment (karthikeyann) [inline python/pylibcudf/examples/regex_search.py]: MEDIUM: `apply_boolean_mask` is deprecated, and its `FutureWarning` is promoted to an error by the focused test configuration. Use the replacement API.

```suggestion
    selected = plc.stream_compaction.apply_retention_mask(
```

File: python/pylibcudf/tests/test_regex_search_example.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+    regex_search.configure_gds(True)
+    assert regex_search.os.environ["KVIKIO_COMPAT_MODE"] == "OFF"
+
+
+def test_native_gds_fallback_is_rejected(monkeypatch):
+    """Reject GDS mode when native GDS is unavailable."""
+    cufile_driver = type(
+        "CuFileDriver", (), {"get": staticmethod(lambda name: False)}
+    )
```
Comment (karthikeyann) [inline python/pylibcudf/tests/test_regex_search_example.py:118]: Update the fake driver to model initialization and ensure property access cannot happen first.

```suggestion
    initialized = False

    def initialize():
        nonlocal initialized
        initialized = True

    def get(name):
        assert initialized
        return False

    cufile_driver = type(
        "CuFileDriver",
        (),
        {"initialize": staticmethod(initialize), "get": staticmethod(get)},
    )
```

File: python/pylibcudf/tests/test_regex_search_example.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    with pytest.raises(RuntimeError, match="native GDS is unavailable"):
+        regex_search._require_native_gds()
+
+
+def test_native_gds_requires_fallback_to_be_disabled(monkeypatch):
+    """Reject GDS mode while cuFile compatibility fallback is allowed."""
+    settings = {"is_gds_available": True, "allow_compat_mode": True}
+    cufile_driver = type(
+        "CuFileDriver", (), {"get": staticmethod(settings.__getitem__)}
+    )
```
Comment (karthikeyann) [inline python/pylibcudf/tests/test_regex_search_example.py:143]: Apply the same initialization-order check to the compatibility-fallback case.

```suggestion
    settings = {"is_gds_available": True, "allow_compat_mode": True}
    initialized = False

    def initialize():
        nonlocal initialized
        initialized = True

    def get(name):
        assert initialized
        return settings[name]

    cufile_driver = type(
        "CuFileDriver",
        (),
        {"initialize": staticmethod(initialize), "get": staticmethod(get)},
    )
```

**Review by karthikeyann — COMMENTED** [review]
Native GDS testing exposed two small issues. These inline suggestions are the tested local diff. Validation: native GDS available with compatibility fallback disabled; 8 focused tests passed.

**Comment by JayYarlagadda** [comment]
@karthikeyann Thanks for testing this on the B200 and for sharing the detailed results. I applied the two fixes from your tested diff in `c35bc9b3`: cuFile is now initialized before checking its runtime properties, and match counting uses `apply_retention_mask`.

The focused suite passes all 8 tests with a current pylibcudf nightly. I also verified the GDS-off path and confirmed that a host without native GDS correctly rejects cuFile compatibility fallback instead of reporting it as native GDS.

File: python/pylibcudf/examples/regex_search.py (hunk 3 above)
Comment (JayYarlagadda) [reply python/pylibcudf/examples/regex_search.py:76]: Fixed in `c35bc9b3`. `_require_native_gds()` now calls `cufile_driver.initialize()` before reading `is_gds_available` or `allow_compat_mode`.

File: python/pylibcudf/examples/regex_search.py (hunk 4 above)
Comment (JayYarlagadda) [reply python/pylibcudf/examples/regex_search.py]: Updated in `c35bc9b3`. Match counting now uses `apply_retention_mask`, and the focused suite passes all 8 tests with the current pylibcudf API.

File: python/pylibcudf/tests/test_regex_search_example.py (hunk 5 above)
Comment (JayYarlagadda) [reply python/pylibcudf/tests/test_regex_search_example.py:118]: Added the initialization-aware fake driver as suggested. The test now asserts that initialization happens before any runtime property is read.

File: python/pylibcudf/tests/test_regex_search_example.py (hunk 6 above)
Comment (JayYarlagadda) [reply python/pylibcudf/tests/test_regex_search_example.py:143]: Applied the same initialization-order check to the compatibility-fallback test. Both GDS validation paths now cover this ordering requirement.
